util/oci/client.go

```diff
@@ -0,0 +1,361 @@
+		}
+		return version.String(), nil
+	}
+
+	return revision, nil
+}
+
+func (c *nativeOCIClient) GetTags(ctx context.Context, noCache bool) (*TagsList, error) {
+	indexLock.Lock(c.repoURL)
+	defer indexLock.Unlock(c.repoURL)
+
+	var data []byte
+	if !noCache && c.indexCache != nil {
+		if err := c.indexCache.GetHelmIndex(c.repoURL, &data); err != nil && !errors.Is(err, cache.ErrCacheMiss) {
+			log.Warnf("Failed to load index cache for repo: %s: %s", c.repoLock, err)
+		}
+	}
+
+	tags := &TagsList{}
+	if len(data) == 0 {
+		start := time.Now()
+		err := c.repo.Tags(ctx, "", func(tagsResult []string) error {
+			for _, tag := range tagsResult {
+				// By convention: Change underscore (_) back to plus (+) to get valid SemVer
+				convertedTag := strings.ReplaceAll(tag, "_", "+")
+				tags.Tags = append(tags.Tags, convertedTag)
+			}
+
+			return nil
+		})
+		if err != nil {
+			return nil, fmt.Errorf("failed to get tags: %w", err)
+		}
+		log.WithFields(
+			log.Fields{"seconds": time.Since(start).Seconds(), "repo": c.repoURL},
+		).Info("took to get tags")
+
+		if c.indexCache != nil {
+			if err := c.indexCache.SetHelmIndex(c.repoURL, data); err != nil {
+				log.Warnf("Failed to store tags list cache for repo: %s: %s", c.repoURL, err)
+			}
+		}
+	} else {
+		err := json.Unmarshal(data, tags)
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode tags: %w", err)
+		}
+	}
+
+	return tags, nil
+}
+
+func newTLSConfig(creds Creds) (*tls.Config, error) {
+	tlsConfig := &tls.Config{InsecureSkipVerify: creds.InsecureSkipVerify}
+
+	if creds.CAPath != "" {
+		caData, err := os.ReadFile(creds.CAPath)
+		if err != nil {
+			return nil, err
+		}
+		caCertPool := x509.NewCertPool()
+		caCertPool.AppendCertsFromPEM(caData)
+		tlsConfig.RootCAs = caCertPool
+	}
+
+	// If a client cert & key is provided then configure TLS config accordingly.
+	if len(creds.CertData) > 0 && len(creds.KeyData) > 0 {
+		cert, err := tls.X509KeyPair(creds.CertData, creds.KeyData)
+		if err != nil {
+			return nil, err
+		}
+		tlsConfig.Certificates = []tls.Certificate{cert}
+	}
+	// nolint:staticcheck
+	tlsConfig.BuildNameToCertificate()
+
+	return tlsConfig, nil
+}
+
+func fileExist(filePath string) (bool, error) {
+	if _, err := os.Stat(filePath); err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		} else {
+			return false, err
+		}
+	}
+	return true, nil
+}
+
+type tarFileStore struct {
+	*oci.Store
+	dest    string
+	tempDir string
+}
+
+func newTarFileStore(tempDir, dest string) *tarFileStore {
+	f, _ := oci.New(tempDir)
+	return &tarFileStore{f, dest, tempDir}
+}
+
+func (s *tarFileStore) Push(ctx context.Context, desc v1.Descriptor, content io.Reader) error {
+	if isTar(desc.MediaType) {
+		all, err := io.ReadAll(content)
+		if err != nil {
+			return err
+		}
+		return os.WriteFile(s.dest, all, 0644)
+	} else {
+		err := s.Store.Push(ctx, desc, content)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+func isTar(mediaType string) bool {
```

Comment:
>Is this function more like isTarGz or something? 

Indeed it is for now. I want to extend this so that it handles simple tarballs, as well as validating that the media type is allowed to be used by the repo-server.